### PR TITLE
#606 - Add HPolyhedron

### LINKED ISSUE #606: Add HPolyhedron type
The only difference to `HPolytope` would be that we do not make the implicit assumption that the set is bounded.

### PR DESCRIPTION
**Pull request by mforets**
Closes #606.

### CODE REVIEW

File: src/HPolyhedron.jl [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+### Output
+
+The support vector in the given direction.
+
+### Algorithm
+
+This implementation uses `GLPKSolverLP` as linear programming backend.
+"""
+function σ(d::AbstractVector{N}, P::HPolyhedron{N}) where {N<:Real}
```
Comment (schillic) [inline src/HPolyhedron.jl]: Can you not use `HPoly here?

File: src/HPolyhedron.jl [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    constraints::Vector{LinearConstraint{N}}
+end
+
+# constructor for an HPolyhedron with no constraints
+HPolyhedron{N}() where {N<:Real} = HPolyhedron{N}(Vector{LinearConstraint{N}}())
+
+# constructor for an HPolyhedron with no constraints of type Float64
+HPolyhedron() = HPolyhedron{Float64}()
+
+# constructor for an HPolytope from a simple H-representation
```
Comment (schillic) [inline src/HPolyhedron.jl]: HPolyhedron

File: src/HPolyhedron.jl [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        error("the support vector is undefined because the polyhedron is empty")
+    else
+        return lp.sol
+    end
+end
+
+"""
+    ∈(x::AbstractVector{N}, P::HPoly{N})::Bool where {N<:Real}
+
+Check whether a given point is contained in a polytope in constraint
```
Comment (schillic) [inline src/HPolyhedron.jl]: polyhedron

File: src/HPolyhedron.jl [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        if dot(c.a, x) > c.b
+            return false
+        end
+    end
+    return true
+end
+
+
+# ===========================================
+# HPolyhedron and HPolytope's shared methods
```
Comment (schillic) [inline src/HPolyhedron.jl]: Some of the methods above are also shared.

File: src/HPolyhedron.jl [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+end
+
+"""
+    constraints_list(P::HPoly{N})::Vector{LinearConstraint{N}} where {N<:Real}
+
+Return the list of constraints defining a polyhedron in H-representation.
+
+### Input
+
+- `P` -- polytope in H-representation
```
Comment (schillic) [inline src/HPolyhedron.jl]: polyhedron

File: src/HPolytope.jl [hunk 6]
```diff
@@ -2,8 +2,7 @@ using MathProgBase, GLPKMathProgInterface
 
 export HPolytope,
        addconstraint!,
-       constraints_list,
-       tosimplehrep
+       constraints_list
```
Comment (schillic) [inline src/HPolytope.jl]: `constraints_list` has also been moved.

File: src/HPolyhedron.jl (hunk 4 above)
Comment (mforets) [reply src/HPolyhedron.jl]: thanks, i pulled out all shared methods excepting the conversions

Comment (schillic) [reply src/HPolyhedron.jl]: But this comment still confuses me.

File: src/concrete_intersection.jl [hunk 7]
```diff
@@ -340,12 +340,16 @@ function intersection(P1::VPolytope{N},
...
 function intersection(P1::VPolytope{N},
                       P2::VPolytope{N};
                       backend=default_polyhedra_backend(N),
                       prunefunc=removehredundancy!) where N
-    return intersection(tohrep(P1), P2; backend=backend, prunefunc=prunefunc)
+    Q1 = polyhedron(P1, backend)
+    Q2 = polyhedron(P2, backend)
+    Pint = Polyhedra.intersect(Q1, Q2)
+    prunefunc(Pint)
+    return convert(typeof(P1), Pint)
```
Comment (schillic) [inline src/concrete_intersection.jl]: Actually, you know the type of `P1` now :wink:

File: src/concrete_intersection.jl [hunk 8]
```diff
@@ -340,12 +340,16 @@ function intersection(P1::VPolytope{N},
     return intersection(P2, P1; backend=backend, prunefunc=prunefunc)
 end
 
-# missing case
+# missing case: VRep wiht VRep
```
Comment (schillic) [inline src/concrete_intersection.jl]: with

**Pull request merged**